### ISSUE TITLE
New example showing differences between MPI_COMM_FREE and MPI_COMM_DISCONNECT

### ISSUE THREAD

**Issue by Wee-Free-Scot** [mpi-forum/mpi-issues#536]
# Problem

The usage differences between MPI_COMM_FREE and MPI_COMM_DISCONNECT are implied by the current normative text but an example shows more clearly the range of freedom permitted.

# Proposal

Add an instructive example showing when MPI_COMM_FREE can be used but MPI_COMM_DISCONNECT cannot replace it.

# Changes to the Text

Add example as follows:

``` C
MPI_Init();
MPI_Comm_dup(MPI_COMM_WORLD, &my_comm);
MPI_Comm_rank(my_comm, &my_rank);
if (0==my_rank) {
  MPI_Ssend_init(dest:=1, comm=my_comm, request=&my_request);
} else {
  MPI_Recv_init(source:=0, comm=my_comm, request:=&my_request);
}
MPI_Comm_free(&my_comm); // do NOT replace with MPI_Comm_disconnect
MPI_Barrier(MPI_COMM_WORLD);
for (int loop=0; loop<100; ++loop) {
  MPI_Start(&my_request);
  MPI_Wait(&my_request, MPI_STATUS_IGNORE);
}
MPI_Request_free(&my_request); // my_comm will not be destroyed before here
MPI_Finalize();
```

# Impact on Implementations

None (assuming they embody the implications of the existing text).

# Impact on Users

None (assuming they understand the implications of the existing text).

# References and Pull Requests

PR TBD

**Comment by correaa** [comment]
For completeness, what should happen if you replace `_free` with `_disconnect`? The program should hang?

**Comment by Wee-Free-Scot** [comment]
> For completeness, what should happen if you replace `_free` with `_disconnect`? The program should hang?

It is not stated explicitly, but I would expect that such a program is erroneous. A high quality implementation of MPI might raise an error on the first usage of the communicator after its disconnection, but a valid MPI implementation might set the datacenter on fire.

**Comment by Wee-Free-Scot** [comment]
I could make an argument that the error code/class that should be used for this situation is MPI_ERR_PROC_FAILED, at least in the case of ULFM, because the link to the other MPI process has been destroyed, no alternative link can serve the same purpose (even though communication may still be possible between the OS processes that host the MPI processes, the specific context of that destroyed communicator is no longer valid), and all “undeliverable” faults are masked as MPI_ERR_PROC_FAILED in ULFM.

@abouteiller we should consider this as another edge-case for the "what happens if?" text about which error handler is used for strange situations.

**Comment by raffenet** [comment]
> ```c
> MPI_Init();
> MPI_Comm_dup(MPI_COMM_WORLD, &my_comm);
> MPI_Comm_rank(my_comm, &my_rank);
> if (0==my_rank) {
>   MPI_Ssend_init(dest:=1, comm=my_comm, request=&my_request);
> } else {
>   MPI_Recv_init(source:=0, comm=my_comm, request:=&my_request);
> }
> MPI_Comm_free(&my_comm); // do NOT replace with MPI_Comm_disconnect
> MPI_Barrier(MPI_COMM_WORLD);
> for (int loop=0; loop<100; ++loop) {
>   MPI_Start(&my_request);
>   MPI_Wait(&my_request, MPI_STATUS_IGNORE);
> }
> MPI_Request_free(&my_request); // my_comm will not be destroyed before here
> MPI_Finalize();
> ```

MPICH will hang. Confirmed in testing.

**Comment by raffenet** [comment]
> For completeness, what should happen if you replace `_free` with `_disconnect`? The program should hang?

MPICH will hang. Confirmed in testing.

**Comment by correaa** [comment]
I don't know if we are in the same page but I think hanging out (with `_DISCONNECT`) is a perfectly valid instance of UB.

**Comment by devreal** [comment]
This example definitely needs surrounding text explaining why I should not replace `MPI_Comm_free` with `MPI_Comm_disconnect`. I tried to figure out why just by reading the description of the two functions and found the text for `MPI_Comm_disconnect` to be somewhat inconsistent:

> `MPI_COMM_DISCONNECT` may be called only if all communication is complete and matched, so that buffered data can be delivered to its destination. This requirement is the same as for `MPI_FINALIZE`.

The "complete" implies that there may not be any incomplete communication. So that supports the comment in the example code. The "matched" part would then be unnecessary (communication has first to be matched and only then can complete, except for RMA where there is no matching).

However, the reference to `MPI_FINALIZE` is not quite correct because finalize calls for all /initiated/ communication to complete:

> It must locally complete all MPI operations that it initiated and must execute matching calls needed to complete MPI communications initiated by other processes.

Initialized persistent requests are not initiated (since they are not in the starting stage yet). 

But then:

> `MPI_COMM_DISCONNECT` has the same action as MPI_COMM_FREE, except that it waits for pending communication to finish internally and enables the guarantee about the behavior of disconnected processes.

This sentence suggests that it is valid to replace `MPI_Comm_free` with `MPI_Comm_disconnect` in the example. I do not consider an initialized persistent request to be pending since it's not started yet. But then, we do not explicitly define the term pending... 

Taking the wording using "complete" in the first part of the description, would the following code be valid?

```C
MPI_Init();
MPI_Comm_dup(MPI_COMM_WORLD, &my_comm);
MPI_Comm_rank(my_comm, &my_rank);
if (0==my_rank) {
  MPI_Ssend_init(dest:=1, comm=my_comm, request=&my_request);
} else {
  MPI_Recv_init(source:=0, comm=my_comm, request:=&my_request);
}
MPI_Barrier(MPI_COMM_WORLD);
for (int loop=0; loop<100; ++loop) {
  MPI_Start(&my_request);
  MPI_Wait(&my_request, MPI_STATUS_IGNORE);
}
MPI_Comm_disconnect(&my_comm); // operations are complete, we're good to go?
MPI_Request_free(&my_request); // my_comm will not be destroyed before here
MPI_Finalize();
```
If this is legal, I am puzzled by the distinction between the initialized and complete states for persistent requests here, since they are both "inactive" according to Figure 2.3.

If the modified example is not legal, the description of `MPI_Comm_disconnect` probably needs to be updated:

> `MPI_COMM_DISCONNECT` may be called only if all communication are able to complete so that buffered data can be delivered to its destination and persistent operations must be in the freeing stage.

The reference to `MPI_FINALIZE` would not be correct as the description of finalize does not mention the requirement to free persistent requests up front.

Bottom line: there seems to be a fair amount of inconsistency here.

[UPDATE] I realized that I used `MPI_Comm_free` in my modified example. It should have been `MPI_Comm_disconnect`. Fixed.[/UPDATE]

**Comment by bosilca** [comment]
In most places we use pending to denote ongoing communications, aka. communications started but not completed. Thus, I agree with @devreal , persistent not-started communications are not pending, and therefore the constraint on MPI_COMM_DISCONNECT should not apply.

**Comment by correaa** [comment]
It makes sense, assuming that request are implemented as to run in a thread for example, if the request is not started there is nothing to really cancel or wait for. This type of request is just sitting in a queue of operations, a queue that is freed with the communicator. 

The remaining question is what if `MPI_request_free` can be called in a request that didn't start yet, if so then it is all good. 

The other question is what happens if the request is asked to start after the communicator was disconnected, I guess UB is the answer in this case.

**Comment by bosilca** [comment]
Persistent requests must always be freed or there is a guarantee of memory leak.

For your second question, what is happening if you start the persistent request after the communicator was disconnected or freed, I would assume this should not be allowed, once freed only pending communications should be allowed to complete. To test this, I used the code above with a small change (move the disconnect above the `MPI_Start`) and, at least, OMPI aborted with an invalid communicator error.

**Comment by Wee-Free-Scot** [comment]
Just catching up with this discussion. Thanks for everyone's comments.

For clarity:
* The comment/instruction `// do NOT replace with MPI_Comm_disconnect` was not meant to indicate a requirement/restriction imposed by the MPI Standard, but rather it is an exhortation to future programmers indicating a point of possible brittleness in the code.

My €0.02 is:
* For disconnect, I *want* it to be bad (erroneous program by preference, but I would accept will raise errors instead) to attempt to start (`MPI_START[_ALL]`), examine {`MPI_REQUEST_GET_STATUS`), test (`MPI_TEST[_*]`, or complete (`MPI_WAIT[_*]`) a request for an operation after the communicator it uses has been disconnected, meaning after `MPI_COMM_DISCONNECT` has been called for that communicator, even if that call has not yet returned. Thus, I *want* it to be the case that `MPI_COMM_DISCONNECT; MPI_START;` would invoke an error handler when `MPI_START` is called. The fault is "{comm_id, proc_rank} is permanently unreachable", which necessitates a failure "undeliverable message", which manifests as an error "MPI process is permanently unreachable", which could be represented by an error class such as `MPI_ERR_UNREACHABLE` (new). This suggests a whole new fault model for MPI, one based on permanently broken virtual communication links rather than fail-stop processes. I have excluded free (`MPI_REQUEST_FREE`) a request because I cannot think of a reason to be draconian about that one: `MPI_COMM_DISCONNECT;MPI_REQUEST_FREE;` should be fine, I guess, because the free is local and just deallocates the request. Extending that, I could countenance permitting the edge-case of using `MPI_REQUEST_GET_STATUS` until it returns `flag==true` (indicating complete but not inactive), then disconnecting the communicator and calling `MPI_TEST` or `MPI_WAIT` (to make the request inactive and, if it is a nonblocking operation, to deallocate the request) and, if it is a persistent request, calling `MPI_REQUEST_FREE` (to deallocate the request).
* For comm_free, I am undecided whether I *want* my code example to be valid or not. There is some (ugly) precedence in MPI for freeing the user's reference to an MPI thing before MPI is done using its reference(s) to that thing (freeing an active point-to-point request). However, any argument for a use-case seems to suffer from the flaw that `MPI_FINALIZE` or `MPI_SESSION_FINALIZE` is needed once "everything is done", which is an opportunity to call `MPI_COMM_FREE` after "everything is done", so a code that calls `MPI_COMM_FREE` early is always unnecessary. OTOH, if `MPI_COMM_FREE` is always delayed, by convention or mandate, until just before `MPI_[SESSION_]FINALIZE` (that is after "everything is done"), what is the point of having *both* `MPI_COMM_FREE` and `MPI_COMM_DISCONNECT`?
* For orthogonalisation, I tend to think of `MPI_COMM_FREE` as `MPI_COMM_IDISCONNECT` with an implicit/internal request that is completed&freed no earlier than when the internal reference count for the communicator hits zero and no later than during the appropriate finalise procedure call, with a lazy implementation that doesn't actually begin any of the disconnection work until the implicit/internal request is being completed. An alternative view is that it simply decrements the reference count of the communicator and, if that makes the reference counter zero, it initiates an internal `MPI_COMM_IDISCONNECT` that is permitted to begin the disconnection work immediately.

**Comment by abouteiller** [comment]
It is erroneous to not complete all requests on the communicator before disconnecting it. That is, the example is -already- erroneous if you substitute COMM_FREE with COMM_DISCONNECT. We don't need new errors to mark the erroneous use of arguments, as we already have ERR_REQUEST, ERR_COMM to indicate use of improper objects.  It is my reading of the standard that the persistent case is 'implicitly' covered: as long as the request is not started, it is not 'pending', and you can disconnect the comm. However, and that text may be missing, it should not be possible to start the request after disconnecting. 

The COMM_FREE case has been valid for years. It is debatable if it is a useful case, but it has been possible for so long that it is impossible to know how many user codes are currently doing that. Note that COMM_FREE does not enforce flushing channels, but the implementation may still be (and in practice will be) capable of tracking the internal completion on operations that are still retaining a freed communicator long before finalize. This is implemented in todays' implementation and is not particularly complex or cost inducing.

**Comment by Wee-Free-Scot** [comment]
> It is erroneous to not complete all requests on the communicator before disconnecting it.

The counter to this has already been made here: no request is active at the time of the MPI_COMM_FREE/MPI_COMM_DISCONNECT. No rule has been broken.

> However, and that text [may be]is missing, it should not be [possible]permitted to start the request after disconnecting.

This is the key ambiguity for the disconnect part.

**Comment by devreal** [comment]
I realized that I used `MPI_Comm_free` in my modified example (https://github.com/mpi-forum/mpi-issues/issues/536#issuecomment-1031674542). It should have been `MPI_Comm_disconnect`. I changed the example.

**Comment by abouteiller** [comment]
I read in the definition of MPI_FINALIZE


> It must locally complete all MPI operations that it initiated and must execute matching calls needed to complete MPI communications initiated by other processes. For example, if the process executed a nonblocking send, it must eventually call MPI_WAIT, MPI_TEST, MPI_REQUEST_FREE, or any derived function; if the process is the target of a send, then it must post the matching receive; if it is part of a group executing a collective operation, then it must have completed its participation in the operation.

Edit: I got convinced I was over-reading the text, because the REQUEST_FREE is possible when using non-persistent requests, and is probably what is referred to here, as used as a 'completing procedure' for the non-persistent operation.

Prior argument that is probably wrong:
> In some clumsy way, this indicates that persistent communication must be freed before MPI_FINALIZE (?) Taking into account that MPI_COMM_DISCONNECT is supposed to have all the same limitations as MPI_FINALIZE for operations on the comm parameter, this would indicate that the example with DISCONNECT is already, and somewhat explicitly, erroneous code.

**Comment by devreal** [comment]
I don't think the definition of `MPI_Finalize` says that I have to free persistent requests before I call `MPI_Finalize`. It says that all communication the process is involved in must be complete before the call. A complete persistent request is by definition, well, complete. There is also no wording that prohibits me from calling `MPI_Request_free` after `MPI_Finalize`. That seems odd to me and maybe it's mentioned somewhere else. In that case, it should be stated explicitly here.

In fact, the text states that:

> The call to MPI_FINALIZE does not free objects created by MPI calls; these objects are freed using MPI_XXX_FREE calls.

This sounds to me (as an unsuspecting user) as if I can call `MPI_XXX_FREE` after I called `MPI_Finalize`...

**Comment by Wee-Free-Scot** [comment]
Picking this up again, there are a couple of things tied together here.
1. calling `MPI_START[_ALL]` with a request that was created using a communicator that has since been disconnected via `MPI_COMM_DISCONNECT` should be explicitly made erroneous -- this requires a new sentence somewhere (probably in the definition text for `MPI_START`, with a reference to/from the definition text for `MPI_COMM_DISCONNECT`) and should be a separate issue from this one, which focuses on the difference between `MPI_COMM_FREE` and `MPI_COMM_DISCONNECT`.
2. the wording for MPI_COMM_DISCONNECT "may be called only if all communication is complete and matched," should be improved either to "may be called only if all communication is complete," ("and matched" is superfluous) or to "may be called only if all communication operations are inactive or enabled," (33% better, once "enabled" is defined as a semantic term) -- this should be a separate issue that may depend on (or be combined into) the issue to define the term "enabled" (TBD).
3. we must decide the analogous rules for MPI_COMM_FREE (any changes should be done in a separate issue): 
  a. can a user initialise (persistent) or initiate (nonblocking) a new operation using a freed communicator? No, because all comm handles are stale.
  b. can a user start an existing (persistent) operation after the communicator has been freed? Currently, yes.
  c. can a user wait/test/request_get_status for an operation (nonblocking or persistent) after the communicator has been freed? Currently, yes.
  d. can a user free a (nonblocking or persistent) request after the communicator has been freed? Currently, yes.
  e. if a user initiates (nonblocking) or starts (persistent) an operation before freeing the communicator, is there a requirement to ensure it is enabled before calling MPI_COMM_FREE (in the same way as there is for MPI_COMM_DISCONNECT)? Currently, no; the enabling call can be after the local/remote call to MPI_COMM_FREE.
4. once the rules for MPI_COMM_FREE are decided/confirmed, we can revisit this issue and design one or more examples that show the differences (if any).

**Comment by correaa** [comment]
You say "can a user start an existing (persistent) operation after the communicator has been freed? Currently, yes."

If so, then that means that at least part of the resources of the communicator are still available somewhere, at least for part of the system.
Is it fair to say then that `MPI_COMM_FREE` then is a misnomer given that it only "marks" the future destruction but doesn't actually return the resources to the system?

Sorry to bring up C++ but coming from C++ I am interested in the deterministic return of resources to the system (for value types), and I think the C++ perspective can help the interpretation of the situation even in the C world.

I think this is another instance in which the dual nature of `MPI_Comm` shows up.
If one interprets it as a value (or unique resource), then `MPI_COMM_FREE` is not really about lifetime but about reference counting.
If one interprets `MPI_Comm` as a pointer (handle) then `MPI_COMM_FREE` is more like the destruction (or `reset`) of a shared pointer (`shared_ptr`).

If that is so, this reinforces the idea that a `mpi3::communicator` entity should use `MPI_COMM_DISCONNECT` for "value" destruction. 
While it is only a "shared-pointer-of-a-communicator"-what should use `MPI_COMM_FREE`.

Which brings yet another subtle question, maybe yet not answered: what can _really_ be done with a handle that was freed ``MPI_COMM_FREE`? if one can do any operation in any other parts of the program then the correspondence with `std::shared_ptr` is almost perfect.
And one can think of having of specialized replacement `std::shared_ptr<mpi3::communicator>`.
If only _some_ operations can be done with it and we want to exploit this, we are in the face of something very difficult to deal with in general.

To be more precise, for the C++ experts: `MPI_COMM_DISCONNECT` should be used to implement a `unique_ptr_of_communicator`? `MPI_COMM_FREE` at best should be used to implement a `shared_ptr_of_communicator`?
If the answer to the second is negative, maybe there is no point in using `MPI_COMM_FREE` except as "a hope for the best, maybe resources will come back to the system" the future state of the program.
In C++ I don't want to deal with this kind of non-determinism because it makes very difficult to reason about the system.

**Comment by Wee-Free-Scot** [comment]
TL;DR; I believe this discussion is interesting for the Languages WG and future MPI, but is out of scope for this specific issue.

@correaa I broadly agree with your assessment of the (mis)match between MPI semantics and C++ semantics.

* An MPI_Comm handle in user space is a direct reference to a communicator object inside MPI. Aliases of this handle do not increase the reference count on the communicator object. There is no way to get multiple different handles that reference the same communicator.
* An MPI_Request handle in user space is an indirect reference to a communicator object inside MPI. Aliases of this handle do not increase the reference count on the communicator object. Initiating/starting multiple MPI operations is a way to obtain multiple different request handles that indirectly reference the same communicator.
* A buffered-mode send where the buffer has not been detached is a doubly-indirect reference to a communicator inside MPI. Aliases of the memory address for the buffer attached to MPI do not increase the reference count on the communicator. There is no way to attach multiple buffers to MPI at the same time. Initiating/starting multiple buffered-mode send operations is a way to obtain multiple doubly-indirect references to the same communicator.

The entire functionality of the communicator is available (in some sense) until the communicator is destroyed. MPI will not destroy a communicator object before all references to it are released. The only guarantees of communicator destruction are MPI_Comm_disconnect and process exit (not just MPI_Finalize, which "cleans up all state" but "does not free objects" doh!).

* MPI_Comm_free decrements the communicator reference count by one and causes all user-space aliases of that communicator handle to become stale.

* MPI_Request_free with an inactive MPI_Request decrements the communicator reference count by one and causes all user-space aliases of that MPI_Request handle to become stale.

* MPI_Request_free with an active MPI_Request removes the request handle from user-space and causes all user-space aliases of that MPI_Request handle to become stale but does NOT decrement the communicator reference count. The operation will continue (c.f. progress); its completion will decrement the communicator reference count by one. The user has no control over, or knowledge of, the precise timing of this action (modulo MPI_T and so on).

* MPI_Buffer_detach does progress until all messages buffered in the buffer have completed locally (meaning MPI is able to, but not actually required to, decrement the communicator reference count for each one as it completes). Once the buffer has been drained of buffered-mode send messages, all the doubly-indirect references to all communicators should have been released (strictly, there is nothing preventing MPI from releasing these doubly-indirect references but there is no way to force MPI to actually release them).

In a high-quality MPI implementation, detaching the buffer (if buffered-mode send has been used), completing all MPI operations (using test/wait), freeing all persistent requests once they are inactive, and freeing the communicator handle (not necessarily in that order) should together be enough to decrement the notional communicator reference count to zero and cause the immediate (strictly, permit the eventual) destruction of the communicator object.

OTOH, MPI_Comm_disconnect does progress until all MPI operations that reference the passed-in communicator complete locally and then destroys the communicator object and the MPI_Comm handle that was passed in, leaving all other references to the communicator dangling.

Neither of these schemes fit well with C++ although both can be implemented using C++. Exposing an actual communicator object to the user is a bad idea because it means exposing an object that has a badly-defined lifetime (from the point-of-view of C++). The user should only ever see the handles, should never use aliasing, and should surmise the effect on a notional internal object and its guessed-at usage of resources.

If a user allows an MPI_Comm handle to go out of scope in a C++ program, should the MPI_Comm destructor call MPI_Comm_free or MPI_Comm_disconnect or neither of these?

Safety:
* Doing nothing is safest -- no possibility of deadlock from making the destructor collective
* Calling MPI_Comm_free is next safest -- collective, but no possibility of making the program erroneous by destroying the communicator too early (in case the user has not complied with "all communication is complete and matched")
* Calling MPI_Comm_disconnect is least safe -- collective and requires the user to have completed all MPI operations that use the communicator

Resources:
* Calling MPI_Comm_disconnect is best -- guarantees the communicator is destroyed and its resources are recovered
* Calling MPI_Comm_free is a compromise -- permits the communicator to be destroyed and its resources to be recovered once it is safe for that to happen
* Doing nothing is worst -- MPI can never know for sure that it is safe for the communicator to be destroyed and its resources to be recovered, what about aliases?

The last point suggests there is a difference between letting a unique_ptr wrapper of a MPI_Comm go out of scope and letting any other type of MPI_Comm reference/variable go out of scope because the unique_ptr guarantees there are no aliases. For other types of reference/variable, C++ would need to manage a reference count for the handle that is separate to the reference count for the object for which it is a handle and only call MPI_Comm_free/MPI_Comm_disconnect once the handle reference count hits zero.

**Comment by VictorEijkhout** [comment]
On , 2022Jul6, at 07:23, Dan Holmes ***@***.******@***.***>> wrote:

If a user allows an MPI_Comm handle to go out of scope in a C++ program,

If by “handle” you mean “ * struct internal_comm_thing” then the C++ equivalent is a shared_ptr and free’ing and such only happens when the reference count runs down.

I think we need to set terms carefully. Currently in C

MPI_Comm foocom = MPI_COMM_WORLD;

Looks like a datatype, but is an opaque handle, whereas in C++

MPI_Comm foocom = MPI.comm_world();

Is probably an actual object. So “foocomm” going out of scope means two very different things in the two languages.

I like the MPL approach where MPI_Comm is an object with a disabled copy constructor, so that you basically only pass it around by reference.

Victor.

**Comment by correaa** [comment]
@Wee-Free-Scot , excellent explanation. It made me understand that, differently to `shared_ptr`, the increment/decrement of the reference counting happens only to indirect means, by creating/freeing requests and not, for example, by copying somehw the `shared_ptr`. It is possible to build a semantic `shared_ptr`, but it will have two overlapping counts. One over the "value" and one over the "internal" count. The language of `shared_ptr` assumes that there is an actual object behind (as @VictorEijkhout ) says but it doesn't matter for `shared_ptr` whether it is copiable or not.  I agree with @VictorEijkhout that, if there is an actual object, it shouldn't be copyable at this stage, but I don't know if it is a fundamental limitation or a quirk in the specification. I also disabled the copy constructor and copy assignment, but I also think that it cannot be the end of the story given that communicators can be duplicated. We can continue the discussion in the Slack MPI C++ channel.

**Comment by Wee-Free-Scot** [comment]
I think this is unlikely to make it into MPI-4.1 due to lack of bandwidth, but I'd like to pick this up again in the MPI-5.0 timeframe. I'll change the labels appropriately and move this into the MPI-5.0 project board.
